### PR TITLE
[SBD] Implement NetAPP ONTAP SAN Driver Support

### LINKED ISSUE #1047: Add NetAPP ONTAP Driver Support
**Is this a BUG REPORT or FEATURE REQUEST?**:

> /kind feature

**What happened**:
As a storage management, it's vital for OpenSDS Controller project to grow its southbound eco-system. Currently we have lvm, ceph etc. as the backend drivers.

**What you expected to happen**:
We are trying to add NetAPP Storage driver to provide support for NetAPP ONTAP storage

### PR DESCRIPTION
**Pull request by himanshuvar**
**What this PR does / why we need it**: As a storage management, it's vital for OpenSDS Controller project to grow its southbound eco-system. Currently we have lvm, ceph etc. as the backend drivers.
This PR is to address implementing NetApp ONTAP Driver Adapter for OpenSDS SBD ecosystem.
Add NetAPP Storage driver to provide support for NetAPP ONTAP storage.

**Which issue this PR fixes** : fixes #1047 

**Special notes for your reviewer**:
To Test the NetApp ONTAP Driver, please do the followings.
1. Copy examples/driver/netapp_ontap_san.yaml to /etc/opensds/driver/ontap.yaml
2. Provide the ONTAP cluster Information in the yaml configuration.
3. Update /etc/opensds/opensds.conf to enable ontap backend.
4. Restart the dock

Tested with ONTAP Select 9.5.

**Release note**:

```
1. Support for NetApp ONTAP SAN Storage Driver Support.
2. Tested with ONTAP Select 9.5 Simulator.
3. Support API Adapters for NetApp CSI project i.e. Trident.
```
Testing is done for the following Operations:

1. Volume Create, List, Get, Delete, Extend
2. Host Create, List, Get, Delete
3. Volume Attachment Create, List, Get, Delete
4. Volume Snapshot Create, List, Get, Delete
5. Create Volume from Snapshot

Please find the following **Output Snapshots and Test Report**

**Test Report**

[NetApp ONTAP Driver Test Report.xlsx](https://github.com/opensds/opensds/files/3978377/NetApp.ONTAP.Driver.Test.Report.xlsx)

### CODE REVIEW

File: contrib/drivers/netapp/ontap/ontap.go [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+		p = defaultConfPath
+	}
+	if _, err := Parse(d.conf, p); err != nil {
+		return err
+	}
+
+	defer func() {
+		if r := recover(); r != nil {
+			log.Errorf("Unable to instantiate backend")
+		}
```
Comment (wisererik) [inline contrib/drivers/netapp/ontap/ontap.go]: please follow https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: contrib/drivers/netapp/ontap/ontap.go [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+		ManagementLIF: d.conf.ManagementLIF,
+		DataLIF:       d.conf.DataLIF,
+		IgroupName:    d.conf.IgroupName,
+		SVM:           d.conf.Svm,
+		Username:      d.conf.Username,
+		Password:      d.conf.Password,
+	}
+	marshaledJSON, err := json.Marshal(config)
+	if err != nil {
+		log.Fatal("Unable to marshal ONTAP config:  ", err)
```
Comment (wisererik) [inline contrib/drivers/netapp/ontap/ontap.go]: same here

File: contrib/drivers/netapp/ontap/ontap.go [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+		return err
+	}
+
+	d.sanStorageDriver = &ontap.SANStorageDriver{
+		Config: *config,
+	}
+
+	// Initialize the driver.
+	if err = d.sanStorageDriver.Initialize("csi", configJSON, commonConfig); err != nil {
+		log.Errorf("Could not initialize storage driver.")
```
Comment (wisererik) [inline contrib/drivers/netapp/ontap/ontap.go]: same here

File: contrib/drivers/netapp/ontap/ontap.go [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+		err = fmt.Errorf("input failed validation: %v", err)
+		return err
+	}
+
+	d.sanStorageDriver = &ontap.SANStorageDriver{
+		Config: *config,
+	}
+
+	// Initialize the driver.
+	if err = d.sanStorageDriver.Initialize("csi", configJSON, commonConfig); err != nil {
```
Comment (wisererik) [inline contrib/drivers/netapp/ontap/ontap.go]: why need csi here, can it be changed?

File: contrib/drivers/netapp/ontap/ontap.go [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+	storagePool := &storage.Pool{
+		Name:               opt.GetPoolName(),
+		StorageClasses:     make([]string, 0),
+		Attributes:         make(map[string]sa.Offer),
+		InternalAttributes: make(map[string]string),
+	}
+
+	err = d.sanStorageDriver.Create(volConfig, storagePool, make(map[string]sa.Request))
+	if err != nil {
+		log.Info("create volume failed")
```
Comment (wisererik) [inline contrib/drivers/netapp/ontap/ontap.go]: please add err into log

File: contrib/drivers/netapp/ontap/ontap.go [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+		return nil, err
+	}
+	log.Info("create volume success")
```
Comment (wisererik) [inline contrib/drivers/netapp/ontap/ontap.go]: it's better add volume name into log here

File: contrib/drivers/netapp/ontap/ontap.go [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+	var snapName = getSnapshotName(opt.GetSnapshotId())
+	var volName = opt.GetMetadata()["volume"]
+	var name = getVolumeName(opt.GetId())
+
+	volConfig := d.GetVolumeConfig(name, opt.GetSize())
+	volConfig.CloneSourceVolumeInternal = volName
+	volConfig.CloneSourceSnapshot = volName
+	volConfig.CloneSourceSnapshot = snapName
+
+	err := d.sanStorageDriver.CreateClone(volConfig)
```
Comment (wisererik) [inline contrib/drivers/netapp/ontap/ontap.go]: didn't hand err here

File: contrib/drivers/netapp/ontap/ontap.go [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+
+	lunPath := lunPath(name)
+
+	// Get LUN Serial Number
+	lunSerialNumber, err := d.getLunSerialNumber(lunPath)
```
Comment (wisererik) [inline contrib/drivers/netapp/ontap/ontap.go]: same here

File: contrib/drivers/netapp/ontap/ontap.go [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+func (d *Driver) PullVolume(volId string) (*model.VolumeSpec, error) {
+
+	return nil, &model.NotImplementError{"method PullVolume has not been implemented yet"}
+}
+
+func (d *Driver) DeleteVolume(opt *pb.DeleteVolumeOpts) error {
+	var name = getVolumeName(opt.GetId())
+	err := d.sanStorageDriver.Destroy(name)
+	if err != nil {
+		msg := fmt.Sprintf("Delete Volume (%s) Failed: %v", opt.GetId(), err)
```
Comment (wisererik) [inline contrib/drivers/netapp/ontap/ontap.go]: same here

File: contrib/drivers/netapp/ontap/ontap.go [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+		log.Error(msg)
+		return err
+	}
+	log.Infof("Delete Volume (%s) Success.", opt.GetId())
```
Comment (wisererik) [inline contrib/drivers/netapp/ontap/ontap.go]: successfully

File: contrib/drivers/netapp/ontap/ontap.go [hunk 11]
```diff
@@ -0,0 +1,463 @@
...
+}
+
+// ExtendVolume ...
+func (d *Driver) ExtendVolume(opt *pb.ExtendVolumeOpts) (*model.VolumeSpec, error) {
+	var name = getVolumeName(opt.GetId())
+	volConfig := d.GetVolumeConfig(name, opt.GetSize())
+
+	newSize := uint64(opt.GetSize() * bytesGB)
+	if err := d.sanStorageDriver.Resize(volConfig, newSize); err != nil {
+		log.Errorf("Extend Volume (%s) Failed, error: %v", name, err)
```
Comment (wisererik) [inline contrib/drivers/netapp/ontap/ontap.go]: same here

File: contrib/drivers/netapp/ontap/ontap.go [hunk 12]
```diff
@@ -0,0 +1,463 @@
...
+
+	publishInfo := &utils.VolumePublishInfo{
+		HostIQN:  []string{initiator},
+		HostIP:   []string{hostInfo.GetIp()},
+		HostName: hostName,
+	}
+
+	err := d.sanStorageDriver.Publish(name, publishInfo)
+	if err != nil {
+		msg := fmt.Sprintf("Attachment is Failed: %v", err)
```
Comment (wisererik) [inline contrib/drivers/netapp/ontap/ontap.go]: same here

File: contrib/drivers/netapp/ontap/ontap.go [hunk 13]
```diff
@@ -0,0 +1,463 @@
...
+
+	snapshot, err := d.sanStorageDriver.CreateSnapshot(snapConfig)
+
+	if err != nil {
+		msg := fmt.Sprintf("create snapshot %s (%s) failed: %s", opt.GetName(), opt.GetId(), err)
+		log.Error(msg)
+		return nil, err
+	}
+
+	log.Infof("Create Snapshot %s (%s) Success.", opt.GetName(), opt.GetId())
```
Comment (wisererik) [inline contrib/drivers/netapp/ontap/ontap.go]: same here

File: contrib/drivers/netapp/ontap/ontap.go (hunk 1 above)
Comment (himanshuvar) [reply contrib/drivers/netapp/ontap/ontap.go]: Done, Thank You

File: contrib/drivers/netapp/ontap/ontap.go (hunk 3 above)
Comment (himanshuvar) [reply contrib/drivers/netapp/ontap/ontap.go]: Done, Thank You

File: contrib/drivers/netapp/ontap/ontap.go (hunk 5 above)
Comment (himanshuvar) [reply contrib/drivers/netapp/ontap/ontap.go]: Done, Thank You

File: contrib/drivers/netapp/ontap/ontap.go (hunk 9 above)
Comment (himanshuvar) [reply contrib/drivers/netapp/ontap/ontap.go]: Done, Thank You

File: contrib/drivers/netapp/ontap/ontap.go (hunk 10 above)
Comment (himanshuvar) [reply contrib/drivers/netapp/ontap/ontap.go]: Done, Thank You

File: contrib/drivers/netapp/ontap/ontap.go (hunk 6 above)
Comment (himanshuvar) [reply contrib/drivers/netapp/ontap/ontap.go]: Done, Thank You

File: contrib/drivers/netapp/ontap/ontap.go (hunk 4 above)
Comment (himanshuvar) [reply contrib/drivers/netapp/ontap/ontap.go]: Its the driver context supported by trident project.
https://github.com/NetApp/trident/blob/master/config/config.go#L108

File: contrib/drivers/netapp/ontap/ontap.go (hunk 2 above)
Comment (himanshuvar) [reply contrib/drivers/netapp/ontap/ontap.go]: Done, Thank You

File: contrib/drivers/netapp/ontap/ontap.go (hunk 7 above)
Comment (himanshuvar) [reply contrib/drivers/netapp/ontap/ontap.go]: Done, Thank You

File: contrib/drivers/netapp/ontap/ontap.go (hunk 8 above)
Comment (himanshuvar) [reply contrib/drivers/netapp/ontap/ontap.go]: Done, Thank You

File: contrib/drivers/netapp/ontap/ontap.go (hunk 11 above)
Comment (himanshuvar) [reply contrib/drivers/netapp/ontap/ontap.go]: Done, Thank You

File: contrib/drivers/netapp/ontap/ontap.go (hunk 13 above)
Comment (himanshuvar) [reply contrib/drivers/netapp/ontap/ontap.go]: Done, Thank You

File: contrib/drivers/netapp/ontap/ontap.go (hunk 4 above)
Comment (wisererik) [reply contrib/drivers/netapp/ontap/ontap.go]: ok

File: contrib/drivers/netapp/ontap/ontap.go [hunk 14]
```diff
@@ -0,0 +1,463 @@
...
+	}
+	log.Infof("remove volume snapshot (%s) success", opt.GetId())
+	return nil
+}
+
+func (d *Driver) ListPools() ([]*model.StoragePoolSpec, error) {
+
+	var pools []*model.StoragePoolSpec
+
+	aggregates, err := d.sanStorageDriver.API.VserverGetAggregateNames()
```
Comment (kumarashit) [inline contrib/drivers/netapp/ontap/ontap.go]: This seems to be generic to the sds requiring vserver. Here the file is ontap.go which appears to be generic.

File: contrib/drivers/utils/config/constants.go [hunk 15]
```diff
@@ -34,6 +34,7 @@ const (
 	ManilaDriverType               = "manila"
 	FujitsuEternusDriverType       = "fujitsu_eternus"
 	ChubaofsDriverType             = "chubaofs"
+	NetappOntapSanDriverType       = "ontap"
```
Comment (kumarashit) [inline contrib/drivers/utils/config/constants.go]: If we support NAS for ONTAP, this constant doesn't seem to be generic.

File: pkg/utils/config/config_define.go [hunk 16]
```diff
@@ -99,6 +99,7 @@ type Backends struct {
 	Manila               BackendProperties `conf:"manila"`
 	FujitsuEternus       BackendProperties `conf:"fujitsu_eternus"`
 	Chubaofs             BackendProperties `conf:"chubaofs"`
+	NetappOntapSan       BackendProperties `conf:"ontap"`
```
Comment (kumarashit) [inline pkg/utils/config/config_define.go]: Same here, what if we support Ontap NAS

File: contrib/drivers/netapp/ontap/constants.go [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+	snapshotPrefix    = "opensds_snapshot_"
+	naaPrefix         = "600a0980"
+	KLvPath           = "lunPath"
+	KLvIdFormat       = "NAA"
+	StorageDriverName = "ontap-san"
+	VolumeVersion     = "1"
+	SnapshotVersion   = "1"
+	accessMode        = ""
+	volumeMode        = "Block"
+	bytesGB           = 1073741824
```
[the remark this answers is not in the document]
Comment (himanshuvar) [reply contrib/drivers/netapp/ontap/constants.go]: API takes input in GiB format.
https://netapp-trident.readthedocs.io/en/stable-v19.01/docker/use/backends/ontap_options.html

File: contrib/drivers/netapp/ontap/constants.go [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+
+package ontap
+
+//  default value for driver
+const (
+	defaultConfPath   = "/etc/opensds/driver/ontap.yaml"
+	DefaultAZ         = "default"
+	volumePrefix      = "opensds_"
+	snapshotPrefix    = "opensds_snapshot_"
+	naaPrefix         = "600a0980"
```
[the remark this answers is not in the document]
Comment (himanshuvar) [reply contrib/drivers/netapp/ontap/constants.go]: Yeah I will change it to 600a98000

[the remark this answers is not in the document]
Comment (kumarashit) [reply contrib/drivers/netapp/ontap/constants.go]: Ok. You can put the comment that it's applicable for C-Mode and 7-Mode

File: contrib/drivers/netapp/ontap/constants.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (kumarashit) [reply contrib/drivers/netapp/ontap/constants.go]: Ok

File: contrib/drivers/netapp/ontap/constants.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (himanshuvar) [reply contrib/drivers/netapp/ontap/constants.go]: Ok Done.

**Pull request merged**